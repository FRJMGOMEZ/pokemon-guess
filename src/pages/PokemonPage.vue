
<template>
  <h1 v-if="!pokemon">Espere por favor...</h1>
  <div v-else>
    <h1>Quién es este pokémon?</h1>
    <PokemonPicture :pokemon-id="pokemon?.id" :show-pokemon="showPokemon" />
    <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer($event)" />
    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame()">Nuevo juego</button>
    </template>
  </div>
</template>

<script>
import PokemonOptions from "../components/PokemonOptions.vue";
import PokemonPicture from "@/components/PokemonPicture.vue";
import getPokemonsOptions from "@/helpers/getPokemonsOptions";

getPokemonsOptions();
export default {
  components: {
    PokemonOptions,
    PokemonPicture,
  },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: null,
      showAnswer: false,
      message: "",
    };
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonsOptions();
      this.pokemon = this.pokemonArr[Math.floor(Math.random() * 4)];
    },
    checkAnswer(pokemonId) {
      if (this.pokemon.id === pokemonId) {
        this.showPokemon = true;
        this.showAnswer = true;
        this.message = `Correcto!, ${this.pokemon.name}`;
      }
    },
    newGame() {
      this.mixPokemonArray().then(() => {
        this.showPokemon = false;
        this.showAnswer = false;
        this.message = "";
      });
    },
  },
  mounted() {
    this.mixPokemonArray();
  },
};
</script>

<style>

button{
   cursor: pointer;
  }
</style>

