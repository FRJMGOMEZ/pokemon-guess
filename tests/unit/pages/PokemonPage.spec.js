import { shallowMount, mount } from '@vue/test-utils';
import PokemonPage from '@/pages/PokemonPage';
import { pokemons } from '../mocks/pokemons.mock';

describe('`PokemonPage Component', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage);
    })

    test('debe de hacer match con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    })

    test('debe de llamar el mixPokemonArray al montar', () => {
        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray');
        shallowMount(PokemonPage);
        expect(mixPokemonArraySpy).toHaveBeenCalled();
    })

    test('debe de hacer match con el snapshot, cuando cargan los pokemon', () => {
        const wrapper = mount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: "",
                }
            }
        })
        expect(wrapper.html()).toMatchSnapshot();
    })

    test('debe de mostrar los componentes PokemonPicture y PokemonOptions', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: "",
                }
            }
        })
        const pokemonPicture = wrapper.find('pokemon-picture-stub');
        expect(pokemonPicture.exists()).toBeTruthy();
        expect(pokemonPicture.attributes('pokemonid')).toBe('1');
        const pokemonOptions = wrapper.find('pokemon-options-stub');
        expect(pokemonOptions.exists()).toBeTruthy();
        expect(pokemonOptions.attributes('pokemons')).toBeTruthy();
    })

    test('pruebas con checkAnswer',async()=>{
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: "",
                }
            }
        })
        await wrapper.vm.checkAnswer(1);
        const h2 = wrapper.find('h2');
        expect(h2.exists()).toBeTruthy();
        expect(h2.text()).toBe(`Correcto!, ${pokemons[0].name}`)
        expect(wrapper.vm.showPokemon).toBe(true);   
    })
    

})