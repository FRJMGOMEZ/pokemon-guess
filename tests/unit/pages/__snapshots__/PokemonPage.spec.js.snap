// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`PokemonPage Component debe de hacer match con el snapshot 1`] = `<h1>Espere por favor...</h1>`;

exports[`\`PokemonPage Component debe de hacer match con el snapshot, cuando cargan los pokemon 1`] = `
<div>
  <h1>Quién es este pokémon?</h1>
  <div class="pokemon-container"><img src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg" alt="pokemon" class="hidden-pokemon">
    <!--v-if-->
  </div>
  <div class="options-container">
    <ul>
      <li>Charmander</li>
      <li>Bulbasur</li>
      <li>Squirtel</li>
      <li>PiKachu</li>
    </ul>
  </div>
  <!--v-if-->
</div>
`;
